tide-cors/src/middleware.rs

```diff
@@ -0,0 +1,285 @@
+        self.allow_credentials = match HeaderValue::from_str(&allow_credentials.to_string()) {
+            Ok(header) => Some(header),
+            Err(_) => None,
+        };
+        self
+    }
+
+    /// Set allow_headers and return new CorsMiddleware
+    pub fn allow_headers<T: Into<HeaderValue>>(mut self, headers: T) -> Self {
+        self.allow_headers = headers.into();
+        self
+    }
+
+    /// Set max_age and return new CorsMiddleware
+    pub fn max_age<T: Into<HeaderValue>>(mut self, max_age: T) -> Self {
+        self.max_age = max_age.into();
+        self
+    }
+
+    /// Set allow_methods and return new CorsMiddleware
+    pub fn allow_methods<T: Into<HeaderValue>>(mut self, methods: T) -> Self {
+        self.allow_methods = methods.into();
+        self
+    }
+
+    /// Set allow_origin and return new CorsMiddleware
+    pub fn allow_origin<T: Into<HeaderValue>>(mut self, origin: T) -> Self {
+        self.allow_origin = origin.into();
+        self
+    }
+
+    /// Set expose_headers and return new CorsMiddleware
+    pub fn expose_headers<T: Into<HeaderValue>>(mut self, headers: T) -> Self {
+        self.expose_headers = Some(headers.into());
+        self
+    }
+
+    fn build_preflight_response(&self) -> http::response::Response<Body> {
+        let mut response = http::Response::builder()
+            .status(StatusCode::OK)
+            .header(
+                header::ACCESS_CONTROL_ALLOW_ORIGIN,
+                self.allow_origin.clone(),
+            )
+            .header(
+                header::ACCESS_CONTROL_ALLOW_METHODS,
+                self.allow_methods.clone(),
+            )
+            .header(
+                header::ACCESS_CONTROL_ALLOW_HEADERS,
+                self.allow_headers.clone(),
+            )
+            .header(header::ACCESS_CONTROL_MAX_AGE, self.max_age.clone())
+            .body(Body::empty())
+            .unwrap();
+
+        if let Some(allow_credentials) = self.allow_credentials.clone() {
+            response
+                .headers_mut()
+                .append(header::ACCESS_CONTROL_ALLOW_CREDENTIALS, allow_credentials);
+        }
+
+        if let Some(expose_headers) = self.expose_headers.clone() {
+            response
+                .headers_mut()
+                .append(header::ACCESS_CONTROL_EXPOSE_HEADERS, expose_headers);
+        }
+
+        response
+    }
+}
+
+impl<State: Send + Sync + 'static> Middleware<State> for CorsMiddleware {
+    fn handle<'a>(&'a self, cx: Context<State>, next: Next<'a, State>) -> BoxFuture<'a, Response> {
+        FutureExt::boxed(async move {
+            // Return results immediately upon preflight request
+            if cx.method() == Method::OPTIONS {
+                return self.build_preflight_response();
+            }
+
+            let mut response = next.run(cx).await;
+            let headers = response.headers_mut();
+
+            headers.append(
+                header::ACCESS_CONTROL_ALLOW_ORIGIN,
```

Comment:
Should the server not display `Access-Control-Allow-Origin` only if it matches or if it is `*` and hide if not?